Source/Core/Core/AchievementManager.cpp

```diff
@@ -146,80 +152,208 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
+                DiscIO::PARTITION_NONE));
+            if (success)
+            {
+              filereader_state->position += requested_bytes;
+              return requested_bytes;
+            }
+            else
+            {
+              return static_cast<size_t>(0);
+            }
+          },
+      .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
+  rc_hash_init_custom_filereader(&volume_reader);
   m_queue.EmplaceItem([this, callback] {
-    const auto resolve_hash_response = ResolveHash(this->m_game_hash);
-    if (resolve_hash_response != ResponseType::SUCCESS || m_game_id == 0)
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, ""))
     {
-      callback(resolve_hash_response);
-      INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
-      OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from volume.");
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
       return;
     }
-
-    const auto start_session_response = StartRASession();
-    if (start_session_response != ResponseType::SUCCESS)
     {
-      callback(start_session_response);
-      WARN_LOG_FMT(ACHIEVEMENTS, "Failed to connect to RetroAchievements server.");
-      OSD::AddMessage("Failed to connect to RetroAchievements server.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
-      return;
+      std::lock_guard lg{m_lock};
+      m_game_hash = std::move(new_hash);
     }
+    m_loading_volume.reset();
+    LoadGameSync(callback);
+  });
+}
+
+void AchievementManager::LoadGameSync(const ResponseCallback& callback)
+{
+  u32 new_game_id = 0;
+  Hash current_hash;
+  {
+    std::lock_guard lg{m_lock};
+    current_hash = m_game_hash;
+  }
+  const auto resolve_hash_response = ResolveHash(this->m_game_hash, &new_game_id);
+  if (resolve_hash_response != ResponseType::SUCCESS || new_game_id == 0)
+  {
+    INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
+    OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    DisableRestoreAchievementManager(true);
+    callback(resolve_hash_response);
+    return;
+  }
+  u32 old_game_id;
+  {
+    std::lock_guard lg{m_lock};
+    old_game_id = m_game_id;
+  }
+  if (new_game_id == old_game_id)
+  {
+    INFO_LOG_FMT(ACHIEVEMENTS, "Alternate hash resolved for current game {}.", old_game_id);
+    callback(ResponseType::SUCCESS);
+    return;
+  }
+  else if (m_game_id != 0)
+  {
+    INFO_LOG_FMT(ACHIEVEMENTS, "Swapping game {} for game {}; achievements disabled.", old_game_id,
+                 new_game_id);
+    OSD::AddMessage("Achievements are now disabled. Please close all running games to re-enable.",
```

Comment:
"Close all running games" sounds a bit odd, since only one game can be running at a time... How about "please close emulation to re-enable"?